mlflow/transformers.py

```diff
@@ -0,0 +1,793 @@
+
+
+def _get_engine_type(model):
+    """
+    Determines the underlying execution engine for the model based on the 3 currently supported
+    deep learning framework backends: ``tensorflow``, ``torch``, or ``flax``.
+    """
+    from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel
+
+    for cls in model.__class__.__mro__:
+        if issubclass(cls, TFPreTrainedModel):
+            return "tensorflow"
+        elif issubclass(cls, PreTrainedModel):
+            return "torch"
+        elif issubclass(cls, FlaxPreTrainedModel):
+            return "flax"
+
+
+def _get_base_model_architecture(model_or_pipeline):
+    """
+    Extracts the base model architecture type from a submitted model.
+    """
+    from transformers import Pipeline
+
+    if isinstance(model_or_pipeline, Pipeline):
+        return model_or_pipeline.model.name_or_path
+    else:
+        return model_or_pipeline.name_or_path
+
+
+def _get_instance_type(model, base: bool):
+    """
+    Utility for extracting the saved object type or, if the `base` argument is set to `True`,
+    the base ABC type of the model.
+    """
+    if base:
+        from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel, Pipeline
+
+        def _get_model_base_class(model):
+            pipeline_abc = Pipeline
+            tf_abc = TFPreTrainedModel
+            torch_abc = PreTrainedModel
+            flax_abc = FlaxPreTrainedModel
+            for cls in model.__class__.__mro__:
+                if issubclass(cls, tf_abc):
+                    return tf_abc.__name__
+                elif issubclass(cls, torch_abc):
+                    return torch_abc.__name__
+                elif issubclass(cls, flax_abc):
+                    return flax_abc.__name__
+                elif issubclass(cls, pipeline_abc):
+                    return pipeline_abc.__name__
+
+        return _get_model_base_class(model)
+    else:
+        return model.__class__.__name__
+
+
+def _should_add_pyfunc_to_model(pipeline) -> bool:
+    """
+    Discriminator for determining whether a particular task type and model instance from within
+    a ``Pipeline`` is currently supported for the pyfunc flavor.
+    Currently, the only tasks that are supported are NLP-based tasks wherein a Pipeline consists
+    solely of a ``Tokenizer`` and a pre-trained model.
+    Audio, Image, and Video pipelines can still be logged and used, but are not available for
+    loading as pyfunc.
+    Similarly, esoteric model types (Graph Models, Timeseries Models, and Reinforcement Learning
+    Models) are not permitted for loading as pyfunc.
+    """
+    import transformers
+
+    impermissible_attrs = {"feature_extractor", "image_processor"}
+    impermissible_model_types = (
+        transformers.GraphormerPreTrainedModel,
+        transformers.InformerPreTrainedModel,
+        transformers.TimeSeriesTransformerPreTrainedModel,
+        transformers.DecisionTransformerPreTrainedModel,
```

Comment:
What about custom subclasses of PreTrainedModel?

https://huggingface.co/docs/transformers/custom_models#writing-a-custom-model

Reply:
This is an excellent point. I think the safe thing to do here to make sure that we're only getting valid architectures from a PreTrainedModel instance is to verify that the model architecture is available on the HF repo. As part of registration of a new custom model architecture, HF does the validation that 'everything works'. For private repos, provided that a user logs in to their account (via `huggingface-cli login` and pass `use_auth_token=True`), the repo/model-name resolution should work. 
I'll whip up a quick validation check for this.

Reply:
Looking at this, there are a few ways that this can be done:
1. Use a PreTrainedModel and override the configs (common use case) - provided that it doesn't have a FeatureExtractor or an ImageProcessor, we're good on this count. 
2. Custom vocabulary for a Tokenizer (fairly common) - we support this right now if the user passes in a Pipeline that only contains a model and a tokenizer (and we will restore their vocabulary overrides if they use a pipeline). If that Pipeline has anything other than those two objects, we won't serve with pyfunc.
3. Total custom architecture with custom model (removing layers, using hidden states, adding layers, etc) isn't going to work if they don't upload it to the HF repository (either private or public). 

I think this is ok. The 3rd option, even in their docs, has walkthroughs about doing this and makes it logically follow that this is the intended use case for custom models architectures. 
What do you think?

Reply:
I guess what I'm asking is how do we know if a custom subclass of PreTrainedModel is permissible or not for use with pyfunc?

Reply:
The `impermissible_attrs` will be the deciding factor here. If the pipeline only contains a model and a tokenizer, then we're dealing with a "pure" LLM that only requires the ColSpec-friendly input types. 
When we have FeatureExtractors or ImageProcessors, that's when we get into the "we have no clue what the input type might be" area (at the moment; we'll probably find some clever way to handle this later). 
As far as my testing this morning has taken me, I can't find an instance where a pipeline that has exclusively a Model and a Tokenizer has any input structure that our plans for pyfunc won't handle.

Reply:
Makes sense! Thanks Ben :)